src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/FSharp/FSharpMapConverter.cs

```diff
@@ -37,7 +38,34 @@ protected override void CreateCollection(ref Utf8JsonReader reader, scoped ref R
         internal sealed override bool IsConvertibleCollection => true;
         protected override void ConvertCollection(ref ReadStack state, JsonSerializerOptions options)
         {
-            state.Current.ReturnValue = _mapConstructor((List<Tuple<TKey, TValue>>)state.Current.ReturnValue!);
+            List<Tuple<TKey, TValue>> listToConvert = (List<Tuple<TKey, TValue>>)state.Current.ReturnValue!;
+            TMap map = _mapConstructor(listToConvert);
+            state.Current.ReturnValue = map;
+
+            if (!options.AllowDuplicateProperties)
+            {
+                int totalItemsAdded = listToConvert.Count;
+                int mapCount = 0;
+
+                if (map is ICollection<KeyValuePair<TKey, TValue>> collection)
+                {
+                    mapCount = collection.Count;
+                }
+                else
+                {
+                    Debug.Fail("F# Map is a collection");
+
+                    foreach (KeyValuePair<TKey, TValue> _ in map)
```

Comment:
Why add logic if we're asserting that this case is impossible? The type [does implement ICollection](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-collections-fsharpmap-2.html) so we can just assert that.

Reply:
Can we assume the version of FSharp is recent enough that it implements the interface? It was added a while ago but I'm not sure if net10.0 needs to support very old FSharp.Core versions.

Reply:
I downloaded the [oldest available FSharp.Core NuGet package](https://www.nuget.org/packages/FSharp.Core/3.0.2) (uploaded back in 2014) and can verify that it was implementing it back then. I think we can scrap the fallback logic.